peps/pep-0747.rst

```diff
@@ -14,121 +14,93 @@ Post-History: `19-Apr-2024 <https://discuss.python.org/t/typeform-spelling-for-a
+:ref:`Type expressions <typing:type-expression>` provide a standardized way
+to specify types in the Python type system. When a type expression is
+evaluated at runtime, the resulting *type form object* encodes the information
+supplied in the type expression. This enables a variety of use cases including
+runtime type checking, introspection, and metaprogramming.
 
-This PEP makes no Python grammar changes. Correct usage of
-``TypeExpr[]`` is intended to be enforced only by static and runtime
-type checkers and need not be enforced by Python itself at runtime.
+Such use cases have proliferated, but there is currently no way to accurately
+annotate functions that accept type form objects. Developers are forced to use
+an overly-wide type like ``object``, which makes some use cases impossible and
+generally reduces type safety. This PEP addresses this limitation by
+introducing a new special form ``typing.TypeForm``.
 
+This PEP makes no changes to the Python grammar. ``TypeForm`` is
+intended to be enforced only by type checkers, not by the Python runtime.
 
-.. _motivation:
 
 Motivation
 ==========
 
-The introduction of ``TypeExpr`` allows new kinds of metaprogramming
-functions that operate on type expressions to be type-annotated and
-understood by type checkers.
+A function that operates on type form objects must understand how type
+expression details are encoded in these objects. For example, the expressions
+``int | str``, ``"int | str"``, ``list[int]``, and ``MyTypeAlias`` are all
+valid type expressions, and they evaluate to instances of ``types.UnionType``,
+``builtins.str``, ``types.GenericAlias``, and ``typing.TypeAliasType``,
+respectively.
 
-For example, here is a function that checks whether a value is
-assignable to a variable of a particular type, and if so returns the
-original value:
+There is currently no way to indicate to a type checker that a function accepts
+type form objects and knows how to work with them. ``TypeForm`` addresses this
+limitation. For example, here is a function that checks whether a value is
+assignable to a specified type and returns None if it is not::
 
-::
-
-   def trycast[T](typx: TypeExpr[T], value: object) -> T | None: ...
+   def trycast[T](typx: TypeForm[T], value: object) -> T | None: ...
 
-The use of ``TypeExpr[]`` and the type variable ``T`` enables the return
-type of this function to be influenced by a ``typx`` value passed at
-runtime, which is quite powerful.
+The use of ``TypeForm`` and the type variable ``T`` describes a relationship
+between the type form passed to parameter ``typx`` and the function's
+return type.
 
-Here is another function that checks whether a value is assignable to a
-variable of a particular type, and if so returns ``True`` (as a special
-``TypeIs[]`` bool [#TypeIsPep]_):
+``TypeForm`` can also be used with `TypeIs <https://typing.readthedocs.io/en/latest/spec/narrowing.html#typeis>`__
```

Comment:
We were missing a reference at typing.readthedocs.io for this one, I added it in https://github.com/python/typing/pull/1852.

```suggestion
``TypeForm`` can also be used with :ref:`typing:typeis`
```